### PR TITLE
GH24241 make Categorical.map transform nans

### LINKED ISSUE #24241: Series.apply on categorical with NaN has wrong behavior
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
>>> print(pd.isna(pd.Series(['A', 'B', pd.np.nan], dtype='category')))
0    False
1    False
2     True
dtype: bool
>>> print(pd.Series(['A', 'B', pd.np.nan], dtype='category').apply(pd.isna))
0    False
1    False
2    False
dtype: object
>>> print(pd.Series(['A', 'A', pd.np.nan], dtype='category').apply(pd.isna))
0    False
1    False
2      NaN
dtype: category
Categories (1, object): [False]
```
#### Problem description

I would expect case 2 (`['A', 'B', pd.np.nan]`) to be either like case 1 or case 3.
I think the correct behavior would be case 1.

Issue #21565 looks similar but not the same

#### Expected Output
```
0    False
1    False
2     True
dtype: bool
```
in all cases.

Behaves as expected if `dtype='category'` is omitted
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Darwin
OS-release: 18.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 4.0.1
pip: 18.1
setuptools: 39.0.1
Cython: 0.29.1
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.1.1
sphinx: None
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 3.7.3
bs4: 4.6.3
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [ ] closes #24241
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Alters `Categorical.map` so that the mapper function is also applied to NaN values. The mentioned bug report brings up the example of calling `apply(pd.isna)` on a categorical series, arguing that the values and dtype of the returned list should be consistent with non-categorical series. This PR makes the values consistent. The discrepancies in dtypes are present in categorical series without NaN's and consistent with documentation.

### CODE REVIEW

File: pandas/core/arrays/categorical.py [hunk 1]
```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
         Index(['first', 'second', nan], dtype='object')
         """
         new_categories = self.categories.map(mapper)
+
         try:
-            return self.from_codes(self._codes.copy(),
-                                   categories=new_categories,
-                                   ordered=self.ordered)
+            if isinstance(mapper, (dict, ABCSeries)):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: why is this a special case? use ``is_dict_like``

File: pandas/core/arrays/categorical.py [hunk 2]
```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
...
+                new_value = mapper[np.nan]
+            else:
+                new_value = mapper(np.nan)
+        except (AttributeError, KeyError, TypeError, ValueError):
+            new_value = np.nan
+
+        try:
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: why are you try/except here?

Comment (jreback) [reply pandas/core/arrays/categorical.py]: how / why can this fail

File: pandas/tests/indexes/test_category.py [hunk 3]
```diff
@@ -311,6 +311,37 @@ def test_map_with_categorical_series(self):
         exp = pd.Index(["odd", "even", "odd", np.nan])
         tm.assert_index_equal(a.map(c), exp)
 
+    @pytest.mark.parametrize('data, f', [[[1, 1, np.nan], pd.isna],
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: write this as
```
@pytest.mark.parametrize(
     'data',
    'f',
      [
         ......
      ]))
.....```

File: pandas/tests/indexes/test_category.py [hunk 4]
```diff
@@ -311,6 +311,37 @@ def test_map_with_categorical_series(self):
...
+        values = pd.Categorical(data)
+        result = values.map(f)
+        if data[1] == 1:
+            expected = pd.Categorical([False, False, True])
+            tm.assert_categorical_equal(result, expected)
+        else:
+            expected = pd.Index([False, False, True])
+            tm.assert_index_equal(result, expected)
+
+    @pytest.mark.parametrize('data, f', [[[1, 1, np.nan], {1: False}],
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: same

File: pandas/tests/indexes/test_category.py [hunk 5]
```diff
@@ -311,6 +311,37 @@ def test_map_with_categorical_series(self):
...
 
+    @pytest.mark.parametrize('data, f', [[[1, 1, np.nan], pd.isna],
+                                         [[1, 2, np.nan], pd.isna],
+                                         [[1, 1, np.nan], {1: False,
+                                                           np.nan: True}],
+                                         [[1, 2, np.nan], {1: False,
+                                                           2: False,
+                                                           np.nan: True}]])
+    def test_map_fill_nan(self, data, f):
+        values = pd.Categorical(data)
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py (hunk 2 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: AttributeError: if mapper calls a method of the element (e.g. `lambda x: x.lower()`)
KeyError: if mapper is a dict without a key for NaN
TypeError: if mapper expects some type other than a float
ValueError: if mapper tries converting float values to ints (e.g. `lambda x: int(x)`)

File: pandas/core/arrays/categorical.py [hunk 6]
```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
...
+            ret = self.from_codes(self._codes.copy(),
+                                  categories=new_categories,
+                                  ordered=self.ordered)
+            if new_value not in ret.categories and any(self._codes == -1):
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: This looks incorrect when `new_value` is NA. NA should never be in the categories, so the `add_categories` doesn't make sense.

**Review by TomAugspurger — COMMENTED** [review]
Can you document the handling of missing values in the docstring?

File: pandas/core/arrays/categorical.py [hunk 7]
```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
...
+
         try:
-            return self.from_codes(self._codes.copy(),
-                                   categories=new_categories,
-                                   ordered=self.ordered)
+            if is_dict_like(mapper):
+                new_value = mapper[np.nan]
+            else:
+                new_value = mapper(np.nan)
+        except (AttributeError, KeyError, TypeError, ValueError):
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: I'm not really comfortable with this. `mapper` is a user-defined function. Consider

```python
def f(x):
    if isnan(x):
        raise TypeError
    ...
```

that TypeError would be swallowed by pandas.

File: pandas/core/arrays/categorical.py (hunk 6 above)
Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py]: I also don't really like the type instability this creates. Now the output dtype depends on whether there are any missing values present in the column.

Should we instead just say that missing values are not handled in the mapping? That you should instead `.fillna(...).map(mapper)` or `.map(mapper).fillna(...)`, depending on how you want to handle missing values? What does handling missing values in `mapper` give us that `.fillna` doesn't?

**Review by TomAugspurger — CHANGES_REQUESTED** [review]

**Comment by JustinZhengBC** [comment]
I decided to go with @TomAugspurger's suggestion to just change the documentation. I did leave in one change in the function, because previously calling `np.take` would cause `NaN` values (represented by -1) to take the last element of `new_categories`.

**Review by TomAugspurger — APPROVED** [review]

File: pandas/core/arrays/categorical.py [hunk 8]
```diff
@@ -1234,6 +1234,9 @@ def map(self, mapper):
                                    categories=new_categories,
                                    ordered=self.ordered)
         except ValueError:
+            if any(self._codes == -1):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you add a comment on what is going on here

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. just a small comment. ping on green.

File: pandas/core/arrays/categorical.py [hunk 9]
```diff
@@ -1234,6 +1234,11 @@ def map(self, mapper):
                                    categories=new_categories,
                                    ordered=self.ordered)
         except ValueError:
+            # NA values are represented in self._codes with -1
+            # np.take causes NA values to take final element in new_categories
+            if any(self._codes == -1):
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Should this use `np.any`? I think `any` will short-circuit, but `np.any` will likely be faster.

**Pull request merged**